esp-hal-common/src/lib.rs

```diff
@@ -202,25 +202,38 @@ mod critical_section_impl {
 
     #[cfg(xtensa)]
     mod xtensa {
+
+        #[cfg(multi_core)]
+        use super::multicore::{LockKind, MULTICORE_LOCK};
+
         unsafe impl critical_section::Impl for super::CriticalSection {
             unsafe fn acquire() -> critical_section::RawRestoreState {
-                let tkn: critical_section::RawRestoreState;
+                let mut tkn: critical_section::RawRestoreState;
                 core::arch::asm!("rsil {0}, 5", out(reg) tkn);
                 #[cfg(multi_core)]
                 {
-                    let guard = super::multicore::MULTICORE_LOCK.lock();
-                    core::mem::forget(guard); // forget it so drop doesn't run
+                    match super::multicore::MULTICORE_LOCK.lock() {
+                        LockKind::Lock => {}
+                        LockKind::Reentry => tkn |= 1 << 31,
```

Comment:
I think we need a comment here that we're using the unused bits of the PS field to smuggle this flag though, and that it _must_ be reset to zero before restoring processor state as the Xtensa TRM states as much (pg 88).

Reply:
You assume I actually checked for unused bits 🙃